### ISSUE TITLE
alternate format for Eloquent's toJson()

### ISSUE THREAD

**Issue by andrisi** [laravel/internals#465]
For some applications scenarios it would be beneficial to have an alternate toJson() which would just output each related table as an array, instead of combining relations into nested structures. This would enable easy client side editing using Vuejs for example, building interfaces based on the data, and would pave the way to allow easy saving of data returned in the same format from the client.

For example I have a book, book_author and author tables, and defined their relations. Instead of getting my books, with nested book_author elements with nested author elements, I would just need a list of book elements, list of book_author elements and list of authors separately.

I looked into it, and Eloquent runs a separate query for each table (so no joins, just filters on ids) - I would simly need theese tables serialised. 

So instead of this:

```json
{
  "id": 2,
  "title": "Litte book of Life",
  "year": 2012,
  "price": 25,
  "book_authors": [{
      "book_id": 2,
      "author_id": 2,
      "role": 1,
      "author": {
        "id": 2,
        "name": "Marie",
        "born": 1970
      }
    }, {
      "book_id": 2,
      "author_id": 5,
      "role": 2,
      "author": {
        "id": 5,
        "name": "Louis",
        "born": 1965
      }
    }
  ]
}
```

If would get something like this:

```json
{
  "book": [{
      "id": 2,
      "title": "Litte book of Life",
      "year": 2012,
      "price": 25
    }
  ],
  "book_author": [{
      "book_id": 2,
      "author_id": 2,
      "role": 1
    },
    {
      "book_id": 2,
      "author_id": 5,
      "role": 2
    }
  ],
  "author": [{
      "id": 2,
      "name": "Marie",
      "born": 1970
    }, 
    {
      "id": 5,
      "name": "Louis",
      "born": 1965
    }
  ]
}
```

**Comment by joshmanders** [comment]
What exactly is the point of having an array with a single item in it, on an entity you know will never have more than one item?

**Comment by andrisi** [comment]
@joshmanders The aim is to have a simple format that works for any case and can easily saved back to a relational database instead of a complicated nested format that client side JS frameworks can only use with after some mingling, and that has no support for saving. And why do you think the "book" element is just for one item. You might query for all books written by a certain person, or in a work in its title.

**Comment by alberto-bottarini** [comment]
Maybe this could be useful to have a lighter json. Imagine a situation where a thousand books have the same author and authors table have twenty or more columns. Returning the "plain relationship" will produce a smaller json.

**Comment by andrisi** [comment]
@alberto-bottarini Yes, that is a key advantage too. Actually a key advantage of any relational database too, and I aim to use the same proven idea here for Eloquent too.

**Comment by andrisi** [comment]
It would also save processing time by eliminating the need to combine separate results into nested structures.

**Comment by salomoni** [comment]
This would really benefit handling Redux/Flux state as it discourages using nested data. 

Currently the convention there is to use [paularmstrong/normalizr](/paularmstrong/normalizr) for transforming deeply indented objects to flat.

**Comment by andrisi** [comment]
@salomoni good to know, thanks. So actually someone made a component to reverse what could be fixed at the root. I hope someone working on Laravel actually reads this issues...

**Comment by salomoni** [comment]
```php
Collection::macro('toFlatJson', function () {
    // return the flattened data
});

Book::with('authors')->get()->toFlatJson();
```

Maybe?

**Comment by andrisi** [comment]
I looked into it, the data is loaded "flat" by Eloquen with a SELECT query for each related table, so this would be lots of wasted energy.

**Comment by andrisi** [comment]
Kind of sad that no one looks at this from the Laravel team... :-( Easy addition and would make a big difference.

**Comment by michaeldyrynda** [comment]
There's been some work done around [API responses](https://github.com/laravel/framework/pull/19449) recently, allowing you to map the responses more easily with the use of transformers. The flexibility is more useful than a single approach as we have currently.

**Comment by andrisi** [comment]
@michaeldyrynda The issue is that the format I'd like to see sent out is the original one, Eloquent does load the relations table-by-table, and only later converts it into the nested structure, so anything that reverses that is extra work - instead of just providing access to the original results.

**Comment by andrisi** [comment]
@taylorotwell Any explanation? You internally do queries table by table so this could be easy. Why force one world view on everyone?
